btree/btree.c

```diff
@@ -4016,313 +4130,733 @@ static struct vkvv_head *vkvv_data(const struct nd *node)
 }
 
+/**
+ * @brief This function will determine if the current key and value
+ *        can be added to the node.
+ */
 static bool vkvv_isfit(struct slot *slot)
 {
-	/**
-	 * @brief This function will determine if the current key and value
-	 *        can be added to the node.
-	 *        It also checks the validity of the record.
-	 *        Modifications will be required in nt_isfit() function
-	 *        declaration as we need to have the sizes of incoming key and
-	 *        value.
-	 *
-	 *        keysize + valsize <= vkvv_space(slot->s_node)
-	 */
+	m0_bcount_t ksize = m0_vec_count(&slot->s_rec.r_key.k_data.ov_vec);
+	m0_bcount_t vsize;
+	uint16_t    dir_entry;
+
+	if (vkvv_level(slot->s_node) == 0) {
+		dir_entry = sizeof(struct dir_rec);
+		vsize     = m0_vec_count(&slot->s_rec.r_val.ov_vec);
+	} else {
+		dir_entry = 0;
+		vsize     = vkvv_get_vspace();
+	}
+	return ksize + vsize + dir_entry <= vkvv_space(slot->s_node);
 }
 
 static void vkvv_done(struct slot *slot, bool modified)
 {
 
 }
 
+/**
+ * @brief This function will determine if the current space for values or keys
+ *        overlaps with the space occupied by directory.
+ */
+static bool vkvv_is_dir_overlap(struct slot *slot)
+{
+	uint32_t          ksize          =
+			m0_vec_count(&slot->s_rec.r_key.k_data.ov_vec);
+	uint32_t          vsize          =
+			m0_vec_count(&slot->s_rec.r_val.ov_vec);
+	struct vkvv_head *h              = vkvv_data(slot->s_node);
+	int               count          = h->vkvv_used;
+
+	void             *start_dir_addr = vkvv_get_dir_addr(slot->s_node);
+	void             *key_addr       = vkvv_key(slot->s_node, count);
+	void             *val_addr       = vkvv_val(slot->s_node, count);
+	void             *end_dir_addr   = start_dir_addr +
+					   sizeof(struct dir_rec) * (count + 2);
+
+	if (key_addr + ksize > start_dir_addr ||
+	    val_addr - vsize < end_dir_addr)
+		return true;
+	else
+		return false;
+}
+
+/**
+ * @brief This function will move directory if the current space for values or
+ *        keys overlaps with the space occupied by directory.
+ */
+static void vkvv_move_dir(struct slot *slot)
+{
+	struct vkvv_head *h              = vkvv_data(slot->s_node);
+	int               count          = h->vkvv_used;
+	uint32_t          ksize          =
+			m0_vec_count(&slot->s_rec.r_key.k_data.ov_vec);
+	uint32_t          vsize          =
+			m0_vec_count(&slot->s_rec.r_val.ov_vec);
+	void             *start_dir_addr = vkvv_get_dir_addr(slot->s_node);
+	uint32_t          dir_size       = sizeof(struct dir_rec) * (count + 2);
+	void             *end_dir_addr   = start_dir_addr + dir_size;
+	void             *key_addr       = vkvv_key(slot->s_node, count);
+	void             *val_addr       = vkvv_val(slot->s_node, count);
+	uint32_t          diff;
+
+	/* Key space overlaps with directory */
+	if (key_addr + ksize > start_dir_addr) {
+		diff = (key_addr + ksize) - start_dir_addr;
+		memmove(start_dir_addr + diff, start_dir_addr, dir_size);
```

Comment:
Do not use memmove directly. Add a function/macro in a file in lib/ folder with a name such as m0_memmove and call this wrapper function/macro here. Make this change in all the other places. Similarly I also see a call to memcpy - please replace that one too accordingly.

Reply:
Updated.